Blog for local storage capacity isolation GA (#36025)
kubernetes/website: content/en/blog/_posts/2022-08-15-local-storage-capacity-isolation-ga.md

---
layout: blog
title: "Kubernetes 1.25: Local Storage Capacity Isolation Reaches GA"

> date: 2022-08-15

Reviewer: Change to 2022-09-19 which is the publish date

Reviewer 2: ```suggestion
date: 2022-09-19
```
(and also please change the path of the file to match)

Reviewer 3: @xing-yang Are you active for this issue?

Reviewer: Yes.  @jingxu97 Can you please address the review comments?

Author: done, also changed the file name.

slug: local-storage-capacity-isolation-ga
---

**Author:** Jing Xu (Google)

> Local ephemeral storage capacity isolation was introduced as a alpha feature in Kubernetes 1.7 and it went beta in 1.9 and with Kubernetes 1.25 we are excited to announce general availability(GA) of this feature.

Reviewer 4: ```suggestion
Local ephemeral storage capacity isolation was introduced as an alpha feature in Kubernetes 1.7 and it went beta in 1.9. With Kubernetes 1.25, we are excited to announce the general availability (GA) of this feature.
```

Author: done

> Pods use ephemeral local storage for scratch space, caching, and for logs. The lifetime of local ephemeral storage does not extend beyond the life of the individual pod. It is exposed to pods using the container’s writable layer, logs directory, and EmptyDir volumes. Before this feature available, there have been issues related to the lack of local storage accounting and isolation, such as Pods do not know how much local storage is available and cannot request guaranteed local storage. Local storage is a best effort resource and pods can be evicted due to other pods filling the local storage.

Reviewer 4: I'm not sure if EmptyDir needs to be encapsulated in a codeblock. I did it. 
```suggestion
Pods use ephemeral local storage for scratch space, caching, and logs. The lifetime of local ephemeral storage does not extend beyond the life of the individual pod. It is exposed to pods using the container’s writable layer, logs directory, and `EmptyDir` volumes. Before this feature was introduced, there were issues related to the lack of local storage accounting and isolation, such as Pods not knowing how much local storage is available and being unable to request guaranteed local storage. Local storage is a best-effort resource and pods can be evicted due to other pods filling the local storage.
```

Author: updated

> The local storage capacity isolation feature allows users to manage local ephemeral storage in the same way as managing CPU and memory. It provides support for capacity isolation of shared storage between pods, such that a pod can be hard limited in its consumption of shared resources by evicting Pods if its consumption of shared storage exceeds that limit. It also allows setting ephemeral storage request for resource reservation. The limits and requests for shared ephemeral-storage are similar to those for memory and CPU consumption.

Reviewer 4: We should link the feature to the concepts docs, if possible or if we have a feature doc, even better.

```suggestion
The [local storage capacity isolation feature](/docs/concepts/configuration/manage-resources-containers/#local-ephemeral-storage) allows users to manage local ephemeral storage in the same way as managing CPU and memory. It provides support for capacity isolation of shared storage between pods, such that a pod can be hard limited in its consumption of shared resources by evicting Pods if its consumption of shared storage exceeds that limit. It also allows setting ephemeral storage requests for resource reservation. The limits and requests for shared `ephemeral-storage` are similar to those for memory and CPU consumption.
```

### How to use local storage capacity isolation

> A typical configuration for local ephemeral storage is to place all different kinds of ephemeral local data (emptyDir volumes, writeable layers, container images, logs) into one filesystem. Typically, both /var/lib/kubelet and /var/log are on the system root filesystem. If users configure the local storage in different ways, kubelet might not be able to correctly measure disk usage and use this feature.

Reviewer 4: ```suggestion
A typical configuration for local ephemeral storage is to place all different kinds of ephemeral local data (emptyDir volumes, writeable layers, container images, logs) into one filesystem. Typically, both /var/lib/kubelet and /var/log are on the system's root filesystem. If users configure the local storage in different ways, kubelet might not be able to correctly measure disk usage and use this feature.
```

#### Setting requests and limits for local ephemeral storage
You can specify `ephemeral-storage` for managing local ephemeral storage. Each container of a Pod can specify either or both of the following:

* `spec.containers[].resources.limits.ephemeral-storage`
* `spec.containers[].resources.requests.ephemeral-storage`

> In the following example, the Pod has two containers. The first container has a request of 2GiB of local ephemeral storage and a limit of 5GiB. The second container requests 6GiB of local storage. Therefore, the Pod has a request of total 10GiB (8GiB+2GiB) of local ephemeral storage, and a limit of 6GiB of local ephemeral storage. It also sets emptyDir sizeLimit to 5GiB. With this setting in pod spec, it will affect how scheulder makes a decision on scheduling pods and also how kubelet evict pods.

Reviewer 4: I'm not sure of the math here. I could be wrong, conceptually, but if the first container requests 2 GiB & the second container requests 6 GiB shouldn't the total requested be 8 GiB? We either need to change the example or the math.

```suggestion
In the following example, the Pod has two containers. The first container has a request of 2GiB of local ephemeral storage and a limit of 5GiB. The second container requests 6GiB of local storage. Therefore, the Pod requests a total of 10GiB (8GiB+2GiB) of local ephemeral storage and enforces a limit of 6GiB of local ephemeral storage. It also sets emptyDir sizeLimit to 5GiB. With this setting in pod spec, it will affect how scheulder makes a decision on scheduling pods and also how kubelet evict pods.
```

summing the limits for the containers in that Pod. In this case, if the sum of the local ephemeral storage usage from all containers and also the Pod's `emptyDir`volumes exceeds the overall Pod storage limit (12GiB), then the kubelet also marks the Pod for eviction. 

Last, in this example, emptyDir volume sets its sizeLimit to 5Gi. It means that if this pod's emptyDir used up more local storage than 5GiB, the pod will be evicted from the node.

#### Setting resource quota and limitRange for local ephemeral storage

This feature adds two more resource quotas for storage. The request and limit set constraints on the total requests/limits of all containers’ in a namespace.
* requests.ephemeral-storage
* limits.ephemeral-storage

```
apiVersion: v1
kind: ResourceQuota
metadata:
  name: storage-resources
spec:
  hard:
    requests.ephemeral-storage: "10Gi"
    limits.ephemeral-storage: "20Gi"
```

Similar to CPU and memory, admin could use LimitRange to set default container’s local storage request/limit, and/or minimum/maximum resource constraints for a namespace.

> Also, ephemeral-storage may be specified to reserve for kubelet or system. example, --system-reserved=[cpu=100m][,][memory=100Mi][,][ephemeral-storage=1Gi][,][pid=1000] --kube-reserved=[cpu=100m][,][memory=100Mi][,][ephemeral-storage=1Gi][,][pid=1000]

Reviewer: How does this "reserve" work together with those requests/limits mentioned above?  Can you add some explanation in the blog?

Author: added more information there. PTAL

### How do I get involved? 

This project, like all of Kubernetes, is the result of hard work by many contributors from diverse backgrounds working together.

> We offer a huge thank you to all the contributors in Kubernetes Storage SIG and CSI community who helped review the design and implementation of the project, including but not limited to the following:</p><ul><li>Benjamin Elder (<a href=https://github.com/BenTheElder>BenTheElder</a>)</li><li>Michelle Au (<a href=https://github.com/msau42>msau42</a>)</li><li>Tim Hockin (<a href=https://github.com/thockin>thockin</a>)</li><li>Jordan Liggitt (<a href=https://github.com/liggitt>liggitt</a>)</li><li>Xing Yang (<a href=https://github.com/xing-yang>xing-yang</a>)</li>

Reviewer: Add a hyper link to https://github.com/kubernetes/community/tree/master/sig-storage?

> We offer a huge thank you to all the contributors in Kubernetes Storage SIG and CSI community who helped review the design and implementation of the project, including but not limited to the following:</p><ul><li>Benjamin Elder (<a href=https://github.com/BenTheElder>BenTheElder</a>)</li><li>Michelle Au (<a href=https://github.com/msau42>msau42</a>)</li><li>Tim Hockin (<a href=https://github.com/thockin>thockin</a>)</li><li>Jordan Liggitt (<a href=https://github.com/liggitt>liggitt</a>)</li><li>Xing Yang (<a href=https://github.com/xing-yang>xing-yang</a>)</li>

Reviewer 2: Eliding the “to” is common in US English. I think that's fine for a blog article.
